Packs/JamfProtect/Integrations/JamfProtectEventCollector/JamfProtectEventCollector.py

```diff
@@ -0,0 +1,615 @@
+            return events, next_page
+        response = client_event_type_func(command_args, next_page)
+        page_info, parsed_data = parse_response(response=response)
+        if next_page := page_info.get("next"):
+            has_next = True
+        events.extend(parsed_data)
+    return events, ""
+
+
+def calculate_fetch_dates(start_date: str, last_run_key: str, last_run: dict, end_date: str = "") -> tuple:
+    """
+    Calculates the start and end dates for fetching events.
+
+    This function takes the start date and end date provided as arguments.
+    If these are not provided, it uses the last run information to calculate the start and end dates.
+    If the last run information is also not available,
+     it uses the current time as the end date and the time one minute before the current time as the start date.
+
+    Args:
+        start_date (str): The start date for fetching events in '%Y-%m-%dT%H:%M:%SZ' format.
+        last_run_key (str): The key to retrieve the last fetch date from the last run dictionary.
+        last_run (dict): A dictionary containing information about the last run.
+        end_date (str, optional): The end date for fetching events in '%Y-%m-%dT%H:%M:%SZ' format. Defaults to "".
+
+    Returns:
+        tuple: A tuple containing two elements:
+            - The start date as a string in the format '%Y-%m-%dT%H:%M:%SZ'.
+            - The end date as a string in the format '%Y-%m-%dT%H:%M:%SZ'.
+    """
+    now_utc_time = get_current_time()
+    # argument > last run > current time
+    start_date = start_date or last_run.get(last_run_key, {}).get('last_fetch') or (
+        (now_utc_time - timedelta(minutes=1)).strftime(DATE_FORMAT))
+    # argument > current time
+    end_date = end_date or now_utc_time.strftime(DATE_FORMAT)
+    return start_date, end_date
+
+
+def fetch_events(client: Client, max_fetch_alerts: int, max_fetch_audits: int, start_date_arg: str = "",
+                 end_date_arg: str = "") -> tuple:
+    """
+    Fetches events from the Jamf Protect API within a specified date range.
+
+    Args:
+        client (Client): An instance of the Client class.
+        max_fetch (int): The maximum number of events to fetch.
+        start_date_arg (str, optional): The start date for fetching events.
+        end_date_arg (str, optional): The end date for fetching events.
+
+    Returns:
+        tuple[List[dict], dict]: A tuple containing two elements:
+            - A list of dictionaries. Each dictionary represents an event.
+            - A dictionary with the key 'last_fetch' and its value representing the end date of the fetched events.
+    """
+    last_run = demisto.getLastRun()
+    alert_events, alert_next_run = [], {}
+    audit_events, audit_next_run = [], {}
+    alert_next_page = last_run.get("alert", {}).get("next_page", "")
+    audit_next_page = last_run.get("audit", {}).get("next_page", "")
+
+    if not (alert_next_page or audit_next_page) or alert_next_page:
```

Comment:
1. I aimed to your second comment, if we have a next page for alerts but don't have one for audit, we know that they fetched instantly so no need to start a new cycle for the audit in that case.
2. It is not enough to have only `if not audit_next_page`. If you have pagination for **both** of the types, It will not trigger any of the cycles, because in the alert "if" the `audit_next_page` will be `True` so if the statement will get `False` and the same for the audit "if". Hence, you must add the first statement to each "if". I just did "De Morgan" on the statement, but it can easily be switched with `(alert_next_page and audit_next_page)`